Make notebook native preview sidebar resizable

nemanjaglumac opened the pull request:
This task belongs to the milestone 3 of #40254.

### Description
This PR introduces a resizable native query preview sidebar using a `ResizableBox` component from `react-resize` library.

Te biggest challenge while working on this task was to conditionally show either the resizable sidebar or the regular, full-width native query preview - all based on the current window width. The only solution I came up with was to literally display two distinct wrappers around the `NativeQueryPreviewSidebar` component. One is the resizable box and the other one is HTML 5 `aside`.

This required moving some of the logic from the CSS module directly to this component.

This PR also introduces a new key (`nativePreviewSidebarWidth`) to the redux store's `qb.uiControls` object.

### How to verify
1. Open a notebook view
2. Show SQL sidebar
3. It should initially be set to 428px
4. Try making it smaller - it should not be possible
5. Try resizing it to a bigger size
6. Close the sql preview
7. Open it again
8. Sidebar should've retained the last set width instead of the initial 428px
9. The maximum width of the sidebar is dictated by the min width of the notebook editor (which is 640px)

### Demo
https://github.com/metabase/metabase/assets/31325167/50053cba-3794-4f37-a598-b81248a2e494

--- 
> [!Important]
> I will cover this behavior with E2E tests in the upcoming PR that will persist both the sidebar state and the sidebar width to the backend user setting.

Please note that we've been able to re-enable the crucial E2E test for the CI thanks to changes in this PR. ReplayIO does not support nested media queries in the Chrome version yet. We got rid of the CSS module altogether so there's nothing stoping us from running this test in the CI again.
https://github.com/metabase/metabase/pull/40984/commits/e76e8de857b43654624598cd60ce4f6198a4687f

nemanjaglumac committed:
Display sql preview on sizes less than 1280px

nemanjaglumac committed:
Handle min notebook width using Mantine responsitve style prop

nemanjaglumac committed:
Display the resizable sidebar on sizes >= 1280px

nemanjaglumac committed:
Set sidebar width on resize

nemanjaglumac committed:
Add custom handle

kamilmielnik commented on frontend/src/metabase/query_builder/components/view/View/NotebookContainer/NotebookContainer.tsx:
```diff
@@ -22,12 +26,31 @@ export const NotebookContainer = ({
   ...props
 }: NotebookContainerProps) => {
   const [shouldShowNotebook, setShouldShowNotebook] = useState(isOpen);
+  const [windowWidth, setWindowWidth] = useState(window.innerWidth);
+
+  useEffect(() => {
+    const handleResize = () => {
+      setWindowWidth(window.innerWidth);
+    };
+
+    window.addEventListener("resize", handleResize);
+
+    return () => {
+      window.removeEventListener("resize", handleResize);
+    };
+  }, [windowWidth]);
```
[`useWindowSize`](https://github.com/streamich/react-use/blob/master/docs/useWindowSize.md)

nemanjaglumac committed:
Simplify window resizing with `useWindowSize` hook

nemanjaglumac committed:
Use "white" from the theme colors

nemanjaglumac committed:
Improve sidebar semantics with `aside`

nemanjaglumac committed:
Define the `NativeQueryPreviewSidebar` as `aside`

nemanjaglumac committed:
Use Mantine style props instead of CSS module

nemanjaglumac committed:
Run the E2E test again using regular browser

We removed the nested media query that ReplayIO doesn't support.
Now this test can run using Replay Chromium as all other tests do.

nemanjaglumac committed:
Add some breathing room between the scrollbar and the resizable sidebar

kamilmielnik commented on frontend/src/metabase/query_builder/components/view/NativeQueryPreviewSidebar/NativeQueryPreviewSidebar.tsx:
```diff
@@ -62,7 +62,15 @@ export const NativeQueryPreviewSidebar = (): JSX.Element => {
+      h="100%"
+      bg={color("white")}
+      display="flex"
+      style={{ flexDirection: "column" }}
```
We should use CSS modules instead of inline styles

kamilmielnik commented on frontend/src/metabase/query_builder/components/view/View/NotebookContainer/NotebookContainer.tsx:
```diff
@@ -37,32 +47,89 @@ export const NotebookContainer = ({
-        <aside
-          className={NC.sqlSidebar}
-          data-testid="native-query-preview-sidebar"
+      {isNativePreviewSidebarOpen && windowWidth < 1280 && (
```
It'd be nice to extract the 1280 as a constant. Also `minNotebookWidth`, `minSidebarWidth`, `handleWidth` and 1px border width which is mentioned in a comment (in `Handle`) but deserves to be a constant.

kamilmielnik commented on frontend/src/metabase/query_builder/components/view/View/NotebookContainer/NotebookContainer.tsx:
```diff
@@ -37,32 +47,89 @@ export const NotebookContainer = ({
-        <Box className={NC.main}>
+        <Box
+          miw={{ lg: minNotebookWidth }}
+          style={{ flex: 1, overflowY: "auto" }}
```
CSS Modules :pray:

kamilmielnik commented on frontend/src/metabase/query_builder/components/view/View/NotebookContainer/NotebookContainer.tsx:
```diff
@@ -37,32 +47,89 @@ export const NotebookContainer = ({
+          w={rem(handleWidth)}
+          left={left}
+          bg={darken("border", 0.03)}
+          style={{
+            zIndex: 5,
+            cursor: "ew-resize",
+            borderRadius: rem(8),
+          }}
```
CSS modules :pray:

kamilmielnik commented on frontend/src/metabase/query_builder/components/view/View/NotebookContainer/NotebookContainer.tsx:
```diff
@@ -37,32 +47,89 @@ export const NotebookContainer = ({
+
   const transformStyle = isOpen ? "translateY(0)" : "translateY(-100%)";
 
+  const Handle = forwardRef<HTMLDivElement, Partial<ResizableBoxProps>>(
```
Can we define the handle outside of `NotebookContainer` please ? :pray:

kamilmielnik reviewed (approved):
Works well :+1: 
I left some cosmetic remarks.

nemanjaglumac commented on frontend/src/metabase/query_builder/components/view/View/NotebookContainer/NotebookContainer.tsx:
```diff
@@ -37,32 +47,89 @@ export const NotebookContainer = ({
       style={{
         transform: transformStyle,
         transition: `transform ${delayBeforeNotRenderingNotebook}ms, opacity ${delayBeforeNotRenderingNotebook}ms`,
+        zIndex: 2,
+        overflowY: "hidden",
```
@ranquild told me in another PR to get rid of CSS modules and to move everything to Mantine. Now the question is where do we draw the line?

If something can be expressed via [style props](https://mantine.dev/styles/style-props/), I think we all agree we should use that. But what if only one or two properties are left and they can only be expressed using an inline-style prop?

Should CSS module exist only for that or is `style` ok?

ranquild replied:
@nemanjaglumac there is currently no agreement on the FE. I think it's worth creating a new RFC since this is all new.

nemanjaglumac commented on frontend/src/metabase/query_builder/components/view/NativeQueryPreviewSidebar/NativeQueryPreviewSidebar.tsx:
```diff
@@ -62,7 +62,15 @@ export const NativeQueryPreviewSidebar = (): JSX.Element => {
 
   return (
-    <Flex direction="column" w="100%" h="100%" bg={color("white")}>
+    <Box
```
I saw that but cannot determine the cause :(

kamilmielnik replied:
We've had similar issues during drag & drop and there are hacks (currently broken though) in place to address such cases, see #40733

nemanjaglumac replied:
Is there any point in applying that hack if it's broken?

kamilmielnik replied:
That hack was specific to the drag&drop library. 

My point is that we can implement an equivalent workaround for this case.
